Q1

In [2]:
# process salaries only
import numpy as np
salaries = np.array([40000, 55000, 60000, 120000, 130000, 200000, 45000, 300000, 250000, 85000])
# sort the list
salaries=np.sort(salaries)
salaries

array([ 40000,  45000,  55000,  60000,  85000, 120000, 130000, 200000,
       250000, 300000])

In [5]:
Q1 = salaries[int(len(salaries)/4)]
Q3 = salaries[int(3*len(salaries)/4)]

Q1, Q3

(55000, 200000)

In [6]:
Iqr = Q3 - Q1

Iqr

145000

In [8]:
upper_limit = Q3 + 1.5*Iqr
lower_limit = Q1 - 1.5*Iqr

upper_limit, lower_limit

(417500.0, -162500.0)

In [9]:
salaries.mean()

128500.0

Q2

min-max normalization
$$
v' = \frac{(v - \text{min}_a)}{(\text{max}_a - \text{min}_a)}(\text{new\_max}_a - \text{new\_min}_a) + \text{new\_min}_a
$$

z-score normalization

$$
v' = \frac{(v - \text{mean}_a)}{\text{standard\_deviation}_a}
$$

normalization with decimal scaling

$$
v' = \frac{(v)}{10^j}
$$
where \( j \) is the smallest integer such that $max(|v'|) < 1$.

In [11]:
# min-max normalization
math = np.array([80, 75, 89, 90, 82])
science = np.array([90, 88, 94, 85, 78])
english = np.array([85, 92, 76, 88, 80])
# compute min and max for each column
math_min, math_max = math.min(), math.max()
science_min, science_max = science.min(), science.max()
english_min, english_max = english.min(), english.max()

print('maxes', math_max, science_max, english_max)
print('mins', math_min, science_min, english_min)

# normalize each column
math_normalized = (math - math_min) / (math_max - math_min)
science_normalized = (science - science_min) / (science_max - science_min)
english_normalized = (english - english_min) / (english_max - english_min)

print(math_normalized)
print(science_normalized)
print(english_normalized)

maxes 90 94 92
mins 75 78 76
[0.33333333 0.         0.93333333 1.         0.46666667]
[0.75   0.625  1.     0.4375 0.    ]
[0.5625 1.     0.     0.75   0.25  ]


In [12]:
# z-score normalization
math_mean, math_std = math.mean(), math.std()
science_mean, science_std = science.mean(), science.std()
english_mean, english_std = english.mean(), english.std()

print ('means', math_mean, science_mean, english_mean)
print ('stds', math_std, science_std, english_std)

math_normalized = (math - math_mean) / math_std
science_normalized = (science - science_mean) / science_std
english_normalized = (english - english_mean) / english_std

print(math_normalized)
print(science_normalized)
print(english_normalized)

means 83.2 87.0 84.2
stds 5.635601121442148 5.366563145999495 5.67097875150313
[-0.56781875 -1.45503555  1.02917149  1.20661485 -0.21293203]
[ 0.55901699  0.186339    1.30437299 -0.372678   -1.67705098]
[ 0.14106912  1.37542395 -1.44595851  0.67007834 -0.7406129 ]


In [15]:
# decimal scaling
max_abs_math, max_abs_science, max_abs_english = np.max(np.abs(math)), np.max(np.abs(science)), np.max(np.abs(english))
print('max abs', max_abs_math, max_abs_science, max_abs_english)

j_math, j_science, j_english = np.ceil(np.log10(max_abs_math)).astype(int), np.ceil(np.log10(max_abs_science)).astype(int), np.ceil(np.log10(max_abs_english)).astype(int)

print('j', j_math, j_science, j_english)
print('j is 2 as decimal spaces are 2, for example data points with 986 will be 3')

math_decimal_scaled = math / (10**j_math)
science_decimal_scaled = science / (10**j_science)
english_decimal_scaled = english / (10**j_english)

print(math_decimal_scaled)
print(science_decimal_scaled)
print(english_decimal_scaled)

max abs 90 94 92
j 2 2 2
j is 2 as decimal spaces are 2, for example data points with 986 will be 3
[0.8  0.75 0.89 0.9  0.82]
[0.9  0.88 0.94 0.85 0.78]
[0.85 0.92 0.76 0.88 0.8 ]


$$
\text{Cov}(X, Y) = \frac{\sum (X_i - \bar{X})(Y_i - \bar{Y})}{n}
$$



   $$
   \text{Var}(X) = \frac{\sum (X_i - \bar{X})^2}{n}
   $$
   $$
   r = \frac{\text{Cov}(X, Y)}{\sigma_X \sigma_Y}
  $$



In [ ]:
import numpy as np

# Data
study_hours = np.array([2, 3, 5, 7, 8])
test_scores = np.array([50, 60, 75, 85, 90])

# Mean
mean_study_hours = np.mean(study_hours)
mean_test_scores = np.mean(test_scores)

# Covariance
covariance = np.mean((study_hours - mean_study_hours) * (test_scores - mean_test_scores))

# Standard deviations
std_dev_study_hours = np.std(study_hours, ddof=0)
std_dev_test_scores = np.std(test_scores, ddof=0)

# Correlation Coefficient
correlation_coefficient = covariance / (std_dev_study_hours * std_dev_test_scores)

print('cov, corr',covariance, correlation_coefficient)

cov, corr 34.0 0.9917975374761581


The relationship between study hours and test scores is very strong and positive. As study hours increase, test scores also tend to increase significantly.

### a. Handle the Missing Values

**Steps:**

1. **Calculate the mean of the available test scores:**

   Available test scores are 50, 75, 85, and 95.
$$
   \text{Mean} = \frac{50 + 75 + 85 + 95}{4} = \frac{305}{4} = 76.25
  $$
2. **Replace missing values with the mean:**

   - For Student_ID 2: Replace NULL with 76.25
   - For Student_ID 5: Replace NULL with 76.25

   Updated dataset:

   ```
   Student_ID | Study_Hours | Test_Score
   1         | 2           | 50
   2         | 3           | 76.25
   3         | 5           | 75
   4         | 7           | 85
   5         | 8           | 76.25
   6         | 9           | 95
   ```
### b. Linear regression
$$
b = \frac{\text{Cov}(X, Y)}{\text{Var}(X)}
$$
   $$
   \text{Cov}(X, Y) = \frac{\sum (X_i - \bar{X})(Y_i - \bar{Y})}{n}
   $$

   $$
   \text{Cov}(X, Y) = \frac{(2 - 5.67)(50 - 76.08) + (3 - 5.67)(76.25 - 76.08) + (5 - 5.67)(75 - 76.08) + (7 - 5.67)(85 - 76.08) + (8 - 5.67)(76.25 - 76.08) + (9 - 5.67)(95 - 76.08)}{6}
   $$

   $$
   \text{Cov}(X, Y) = \frac{(-3.67)(-26.08) + (-2.67)(0.17) + (-0.67)(-1.08) + (1.33)(8.92) + (2.33)(0.17) + (3.33)(18.92)}{6}
   $$

   $$
   \text{Cov}(X, Y) = \frac{95.73 + 0.45 + 0.72 + 11.89 + 0.40 + 63.08}{6}
   $$

   $$
   \text{Cov}(X, Y) = \frac{171.27}{6} \approx 28.55
   $$

   Variance of Study_Hours:

   $$
   \text{Var}(X) = \frac{\sum (X_i - \bar{X})^2}{n}
   $$

   $$
   \text{Var}(X) = \frac{(2 - 5.67)^2 + (3 - 5.67)^2 + (5 - 5.67)^2 + (7 - 5.67)^2 + (8 - 5.67)^2 + (9 - 5.67)^2}{6}
   $$

   $$
   \text{Var}(X) = \frac{13.47 + 7.13 + 0.45 + 1.77 + 5.43 + 11.21}{6}
   $$

   $$
   \text{Var}(X) = \frac{39.56}{6} \approx 6.59
   $$

   $$
   \sigma_X^2 = 6.59
   $$


   $$
   b = \frac{\text{Cov}(X, Y)}{\text{Var}(X)} = \frac{28.55}{6.59} \approx 4.33
   $$

   $$
   a = \bar{Y} - b \bar{X}
   $$

   $$
   a = 76.08 - (4.33 \times 5.67) \approx 76.08 - 24.60 \approx 51.48
   $$

   So the regression equation is:

   $$
   Y = 51.48 + 4.33X
   $$

### d. Predict the Test Score for a Student Who Studies 6 Hours

Using the regression equation \( Y = 51.48 + 4.33X \):

   $$
   Y = 51.48 + 4.33 \times 6
   $$

   $$
   Y = 51.48 + 25.98 = 77.46
   $$

   So, the predicted test score for a student who studies 6 hours is approximately 77.46.


In [ ]:
import numpy as np
import pandas as pd

# Data
data = {
    'Study_Hours': np.array([2, 3, 5, 7, 8, 9]),
    'Test_Score': np.array([50, 76.25, 75, 85, 76.25, 95])
}

df = pd.DataFrame(data)

# Means
mean_X = np.mean(df['Study_Hours'])
mean_Y = np.mean(df['Test_Score'])

# Covariance and Variance
cov_XY = np.mean((df['Study_Hours'] - mean_X) * (df['Test_Score'] - mean_Y))
var_X = np.mean((df['Study_Hours'] - mean_X) ** 2)

# Slope and Intercept
b = cov_XY / var_X
a = mean_Y - b * mean_X

# Predicting for Study_Hours = 6
predicted_score = a + b * 6

a, b, predicted_score


(51.57838983050847, 4.353813559322034, 77.70127118644066)